lib/class-wp-theme-json-resolver.php

```diff
@@ -71,6 +71,86 @@ private static function get_from_file( $file_path ) {
 		return $config;
 	}
 
+	/**
+	 * Processes a tree from i18n-theme.json into a liner array
+	 * containing the a translatable path from theme.json and an array
+	 * of properties that are translatable.
+	 *
+	 * @param array $file_structure_partial A part of a theme.json i18n tree.
+	 * @param array $current_path           An array with a path on the theme.json i18n tree.
+	 *
+	 * @return array An array of arrays each one containing a translatable path and an array of properties that are translatable.
+	 */
+	private static function theme_json_i18_file_structure_to_paths( $file_structure_partial, $current_path = array() ) {
+		$result = array();
+		foreach ( $file_structure_partial as $property => $partial_child ) {
+			if ( is_numeric( $property ) ) {
+				return array(
+					array(
+						'path'              => $current_path,
+						'translatable_keys' => $file_structure_partial,
+					),
+				);
+			}
+			$result = array_merge(
+				$result,
+				self::theme_json_i18_file_structure_to_paths( $partial_child, array_merge( $current_path, array( $property ) ) )
+			);
+		}
+		return $result;
+	}
+
+	/**
+	 * Returns a data structure used in theme.json translation.
+	 *
+	 * @return array An array of theme.json paths that are translatable and the keys that are translatable
+	 */
+	private static function get_theme_json_i18n() {
+		static $theme_json_i18n = null;
+		if ( null === $theme_json_i18n ) {
+			$file_structure  = self::get_from_file( __DIR__ . '/experimental-i18n-theme.json' );
+			$theme_json_i18n = self::theme_json_i18_file_structure_to_paths( $file_structure );
+
+		}
+		return $theme_json_i18n;
+	}
+
+	/**
+	 * Translates a theme.json structure.
+	 *
+	 * @param array  $theme_json_structure A theme.json structure that is going to be translatable.
+	 * @param string $domain               Optional. Text domain. Unique identifier for retrieving translated strings.
+	 *                                     Default 'default'.
+	 */
+	private static function apply_theme_json_translations( &$theme_json_structure, $domain = 'default' ) {
```

Comment:
Any specific reason for call by reference vs call by value?

Reply:
The only reason why I am passing by reference instead of returning a newly translated theme.json is to avoid copying the entire tree. I don't think it makes a big difference so if you think it is better to return I can change.